Proposed rechartering (#7)
ietf-wg-masque/wg-materials: charter.md

# MASQUE Charter

Many network topologies lead to situations where transport protocol proxying is beneficial. For example, proxying
enables endpoints to communicate when end-to-end connectivity is not possible or to apply additional encryption where
desirable (such as a VPN). Proxying can also improve client privacy, e.g., by hiding a client's IP address from a
target server.

Proxying technologies such as SOCKS and HTTP(S) CONNECT exist, albeit with their own shortcomings. For example, SOCKS
signalling is not encrypted and HTTP CONNECT is currently limited to TCP. In contrast, HTTP/3 is a viable candidate
protocol for proxying arbitrary traffic, as it provides a single connection with secure connectivity, multiplexed
streams, address migration, and a unified congestion controller. An HTTP/3 datagram construct built on top of QUIC
datagram frames provides for unreliable data transmission and enables transporting UDP and other unreliable flows via a
proxy. Moreover, it avoids introducing potentially redundant or unnecessary recovery mechanisms. Lastly, HTTP supports
an established request/response semantic that can set up and configure flows for different services.

> 

Reviewer: These sentences are redundant with the third paragarph.

Author: Updated in [e06ed3e](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/e06ed3e41b0802294715b3b6bf26c6938828132c)

an established request/response semantic that can set up and configure flows for different services.

The primary goal of this working group is to develop mechanism(s) that allow configuring and concurrently running
multiple proxied stream- and datagram-based flows inside an HTTP connection. These mechanism(s) are collectively called
MASQUE. The group will specify HTTP and/or HTTP/3 extensions to enable this functionality.

> The group will focus on a limited set of client-initiated services: (1) UDP CONNECT and (2) IP proxying.

Reviewer 2: ```suggestion
The group will focus on extensions for a limited set of client-initiated services: (1) UDP CONNECT and (2) IP proxying.
```

Server-initiated services are out of scope.

Exercising the extension points defined by CONNECT-UDP and CONNECT-IP helps make it easier to support new use cases or
accommodate changes in the environment in which these protocols are deployed. As such, the working group will consider

> adopting an extension to these documents when it provides missing functionality or otherwise eases deployability for

Reviewer 3: Is performance a reasonable criterion?  If so, why not include that too?

Author: Great idea, added

use cases, where there are multiple implementation and/or deployment proponents. The initial set of extensions will be
in support of UDP listening and QUIC-aware proxying. Additional extensions that help enable CONNECT-UDP or CONNECT-IP

> services to coordinate or work well in the ecosystem of other services may be considered. Extensions to HTTP Datagrams

Reviewer 3: These two sentences seem to overlap a fair bit.

> As such, the working group will consider adopting an extension to these documents when it provides missing functionality or otherwise eases deployability for use cases, where there are multiple implementation and/or deployment proponents.

And

> Additional extensions that help enable CONNECT-UDP or CONNECT-IP services to coordinate or work well in the ecosystem of other services may be considered. 

Can this be simplified?

Author: Thanks, reworked that paragraph and it is now much more concise

> services to coordinate or work well in the ecosystem of other services may be considered. Extensions to HTTP Datagrams

Reviewer 3: ```suggestion
services to coordinate or work well in the ecosystem of other services may be considered.

Extensions to HTTP Datagrams
```

Author: Split this and reordered some of the later paragraphs

will be coordinated with HTTPBIS. Extensions that solely relate to generic proxying functionality, and are not specific
to the core MASQUE documents, are out of scope.

> The working group will consider fallback to versions of HTTP that operate over TCP as a mitigation to UDP or HTTP/3

Reviewer 2: After re-reading the new charter text, I wonder if this first sentence on fallback to TCP is still relevant. Perhaps it ties into the proposed focus on extensions? I don't think it's harmful to include or keep, but flagging just in case others see an issue with it.

Author: Seems fine in that we can continue to consider it :) 
Agreed that it doesn't seem harmful, at least.

Reviewer 3: I might phrase this differently, perhaps:

> The working group will consider how the protocols it defines might operate over versions of HTTP that use TCP rather than QUIC, for use when QUIC is unavailable.  This might include defining alternative extensions specifically for use in these HTTP versions.

Though I will note that this is already implied by the earlier statements about extensions that improve the ability to operate, deploy, or use MASQUE features.

Author: Nice, updated to use this wording, and moved some of this around too -- thank you!

blocking. Specifying proxy server discovery mechanisms is out of scope. New congestion control and loss recovery
algorithms are also out of scope. However, the working group will consider implications of tunneling protocols with
congestion control and loss recovery over MASQUE proxies, and may issue recommendations accordingly. Finally, multicast

> support is out of scope, though the group may specify extension points that would enable future work on multicast.

Reviewer 3: This is an odd statement to make.  Multicast is either out of scope or not, but this says both.

I understand that there are generic capabilities that the IP proxying stuff might need that would be useful in a multicast setting.  So maybe this should say:

> Multicast is out of scope.  Extensions providing generic capabilities that might include multicast are acceptable on the basis of their application to other use cases.  For example, a generic IP proxy might be used for both unicast and multicast communication.  For such generic capabilities, designs will not explicitly disable multicast, but nor will they include multicast-specific features.

Or something like that.

Author: Makes sense, and thanks for the text! Used it almost verbatim

an established request/response semantic that can set up and configure flows for different services.

The primary goal of this working group is to develop mechanism(s) that allow configuring and concurrently running
multiple proxied stream- and datagram-based flows inside an HTTP connection. These mechanism(s) are collectively called
MASQUE. The group will specify HTTP and/or HTTP/3 extensions to enable this functionality.

> The group will focus on extensions for a limited set of client-initiated services: (1) UDP CONNECT and (2) IP proxying.

Reviewer: I don't think paragraph has been updated enough; the reader certainly wouldn't get a clue from this charter that the WG has already delivered UDP-CONNECT and IP proxying.

Author: Updated in [035ae9d](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/035ae9d40aed3c6b645212a82a910f59adcc7f17)

> Server-initiated services are out of scope.

Reviewer: Some people don't think UDP Listen is an example of a "server-initiated service", but if not, I don't know what is ruled out of scope by this sentence. I suggest we simply strike this.

Reviewer 4: I have a pretty clear idea in my mind of what this tries to accomplish, I'd rather we figure out a way to phrase this instead of striking it. Perhaps something like:
```suggestion
MASQUE is only chartered to work on services that the client requests from the server.
Services that the server initiates without any prompt from the client are out of scope.
```
I think that makes it clear that UDP listen is in scope while still scoping out real work items.

Reviewer: Maybe this is a semantic problem. Do you mean "proxy-initiated services"?

Reviewer 4: Indeed, I meant proxy (or more specifically UDP proxy as defined in RFC 9298 or IP proxy from that draft). We didn't have that term defined back when we wrote the first charter.
```suggestion
MASQUE is only chartered to work on services that the client requests from the proxy.
Services that the proxy initiates without any prompt from the client are out of scope.
```

Reviewer: WFM

or accommodate changes in the environment in which these protocols are deployed. The initial set of extensions will be
in support of UDP listening and QUIC-aware proxying. Additional extensions that provide missing functionality, improve
performance, or otherwise ease deployability for use cases may be adopted where there are multiple implementation
and/or deployment proponents. 

Extensions to HTTP Datagrams will be coordinated with HTTPBIS. Extensions that solely relate to generic proxying
functionality, and are not specific to the core MASQUE documents, are out of scope. 

Specifying proxy server discovery mechanisms is out of scope. New congestion control and loss recovery algorithms are
also out of scope. However, the working group will consider implications of tunneling protocols with congestion control
and loss recovery over MASQUE proxies, and may issue recommendations accordingly. 

The working group will consider how the protocols it defines might operate over versions of HTTP that use TCP rather
than QUIC, for use when QUIC is unavailable. This might include defining alternative extensions specifically for use in
these HTTP versions.

> Multicast is out of scope. Extensions providing generic capabilities that might include multicast are acceptable on the

Reviewer 5: Doesn't hurt to be as specific as possible, don't confuse it with "application-layer multicast" etc

```suggestion
IP multicast is out of scope. Extensions providing generic capabilities that might include IP multicast are acceptable on the
```

an established request/response semantic that can set up and configure flows for different services.

The primary goal of this working group is to develop mechanism(s) that allow configuring and concurrently running
multiple proxied stream- and datagram-based flows inside an HTTP connection. These mechanism(s) are collectively called
MASQUE. The group will specify HTTP and/or HTTP/3 extensions to enable this functionality.

> The group will focus on extensions for a limited set of client-initiated services: (1) UDP CONNECT and (2) IP proxying.
> MASQUE is only chartered to work on services that the client requests from the proxy.

Reviewer 2: ```suggestion
Since the MASQUE working group already delivered protocol solutions for CONNECT-UDP and CONNECT-IP,
the group will now focus on extensions for these two client-initiated services.
```

Reviewer 2: @martinduke please have another look. We would like to close the consensus call and hand you the token to move this forward ASAP!

a requirements document for IP proxying. Once both are complete, the working group will focus on a protocol solution for
IP proxying.

> Exercising the extension points defined by CONNECT-UDP and CONNECT-IP is important to ensure their continued useability.

Reviewer 6: I'm not convinced that "useability" is a word.  Also, if the goal is to ensure that extensions are safe to negotiate (i.e. not prevented by server intolerance), that is quite different from "providing missing functionality" or "enhancing performance".

Reviewer 2: Fair point. I left a suggestion that tries to clarify the intent of bringing extensions into the group.

Author: Good call, updated with Chris's text in [8af81ca](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/8af81ca9873dc5f769a87707af732f24be6455bf)

The working group will consider extensions to these documents that provide missing functionality or enhance performance

> for real world use cases. The initial set of extensions will be in support of UDP listening, QUIC aware proxying, and

Reviewer 5: ```suggestion
for real world use cases. The initial set of extensions will be in support of UDP listening, QUIC-aware proxying, and
```

> datagram priorities. Extensions that help enable CONNECT-UDP or CONNECT-IP services to coordinate or work well in the

Reviewer 6: I am not convinced that we have a sensible use case for datagram priorities, so I would prefer not to name it in the charter.

Reviewer 2: I'm OK folding this into the blanket acceptance clause we have for additional extensions.

Reviewer 4: +1, I'm not hearing anyone saying they want to work on datagram priorities

Reviewer 5: I think the use case exists and I would work on it. But I can live with it not being mentioned in the charter explicitly, and it running through the standard process of seeking adoption as an extension.

Author: Removed, but also comfortable with doing this as clearing the bar laid out in the non-explicitly mentioned set

> ecosystem of other services may be considered. Extensions to HTTP/3 Datagrams will be coordinated with HTTPBIS.

Reviewer 5: HTTP/3 Datagrams or HTTP Datagrams?

> ecosystem of other services may be considered. Extensions to HTTP/3 Datagrams will be coordinated with HTTPBIS.

Reviewer 2: ```suggestion
ecosystem of other services may be considered. Extensions to HTTP Datagrams will be coordinated with HTTPBIS.
```

Extensions that solely relate to generic proxying functionality, and are not specific to the core MASQUE documents, are
out of scope.

> The working group will consider fallback to versions of HTTPS that operate over TCP as a mitigation to UDP or HTTP/3

Reviewer 2: ```suggestion
The working group will consider fallback to versions of HTTP that operate over TCP as a mitigation to UDP or HTTP/3
```

> The working group will consider fallback to versions of HTTPS that operate over TCP as a mitigation to UDP or HTTP/3

Reviewer 7: Versions of HTTP, not HTTPS?

Author: Yeah, this was in there previously as 
> The working group will consider fallback to versions of HTTPS that operate over TCP as a mitigation to UDP or HTTP/3

but seems good to clean up something this small either way

Reviewer 2: I'm OK with "versions of HTTP" for this 👍

Reviewer 4: +1

Author: Updated in [8af81ca](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/8af81ca9873dc5f769a87707af732f24be6455bf)

> blocking. Specifying proxy server discovery mechanisms is also out of scope. Moreover, the working group will consider

Reviewer 2: ```suggestion
blocking. Specifying proxy server discovery mechanisms is out of scope. Moreover, the working group will consider
```

> blocking. Specifying proxy server discovery mechanisms is out of scope. Moreover, the working group will consider
> implications of tunneling protocols with congestion control and loss recovery over MASQUE proxies, and may issue
> recommendations accordingly. New congestion control and loss recovery algorithms are out of scope. Multicast support is
> out of scope. However, the group may specify extension points that would enable future work on multicast.

Reviewer 2: ```suggestion
blocking. Specifying proxy server discovery mechanisms is out of scope. New congestion control and loss recovery
algorithms are also out of scope. However, the working group will consider implications of tunneling protocols
with congestion control and loss recovery over MASQUE proxies, and may issue recommendations accordingly.
Finally, multicast support is out of scope, though the group may specify extension points that would enable
future work on multicast.
```

> out of scope. However, the group may specify extension points that would enable future work on multicast.

Reviewer 5: The content in this paragraph is ok but the presentation makes it hard to digest. Some specific comments

> Specifying proxy server discovery mechanisms is also out of scope.

"Also out of scope" reads weird now that you've moved the multicast stuff later in the paragraph. 

> Moreover, ... 

I'm not sure if this is supposed to be a moreover related to TCP fallback or server discovery. If its the latter that's weird.

* "will consider CC and loss recovery" next sentence "new CC and loss recovery is out of scope". I think I understand what you're pitching at but this could be phrased more clearly to avoid looking like conflict

Reviewer 2: Thanks for flagging this odd language! I sent a suggestion that should hopefully make this more clear.

Author: Thanks! Updated in [8af81ca](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/8af81ca9873dc5f769a87707af732f24be6455bf)

> a requirements document for IP proxying. Once both are complete, the working group will focus on a protocol solution for

Reviewer 4: Since we're rechartering, I would remove the text about our past deliverables - they're no longer our goal/purpose

Reviewer 2: Good point!

Author: Thought about that initially and was trying to keep it a minimal diff, but I'll cut these lines out since we did do them and just leave the rest of the "focus on initial set" since that is a true statement about the group's focus (i.e. server-initiated services remain out of scope).

Author: Removed in [8af81ca](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/8af81ca9873dc5f769a87707af732f24be6455bf)

IP proxying.

> Exercising the extension points defined by CONNECT-UDP and CONNECT-IP is important to ensure their continued useability.
> The working group will consider extensions to these documents that provide missing functionality or enhance performance
> for real world use cases. The initial set of extensions will be in support of UDP listening, QUIC-aware proxying, and

Reviewer 2: ```suggestion
Exercising the extension points defined by CONNECT-UDP and CONNECT-IP helps make it easier to
support new use cases or accommodate changes in the environment in which these protocols are deployed.
As such, the working group will consider extensions to these documents that provide missing
functionality or otherwise ease deployability for use cases with implementation and deployment proponents.
The initial set of extensions will be in support of UDP listening, QUIC-aware proxying, and
```

Author: Also folded in the later discussion, this is much nicer, thank you!

> for real world use cases. The initial set of extensions will be in support of UDP listening, QUIC-aware proxying, and

Reviewer 4: We might want to define what a "real world use case" is more formally now to avoid arguing about it later. It looks like it's used twice in this new charter and is the gating factor for new work so it's bound to come up

Reviewer 2: I think that's a good idea, but I don't really have a great definition for "real world use cases." I'll think on this.

Reviewer 2: Since use cases that are real world are ultimately driven by people who care about them and want to support them, perhaps we can say something like "use cases with implementation and deployment proponents"?

Reviewer 2: (I modified the suggestion to use this phrasing)

Reviewer 5: The extension mechanisms are decently protected from observer meddling. Folks can use and experiment with the extension points without strong coordination from a standards body.

I'm guessing we're mainly interested in adopting work to the WG (on a standards track?) that has sufficient interest from the WG beyond just single proponent or deployment.

Reviewer 2: Right -- hence plural "proponents". Happy to take alternative phrasings if they're more clear.

Reviewer 5: Ah, I think my reading of the multiple plurals in one sentence did not grasp that intent then. How about something like

> As such, the working group will consider adopting an extension to these documents when it provides missing
functionality, or otherwise eases deployability for use cases, where there are multiple implementation and/or deployment proponents.

Reviewer 2: That reads good to me. @ekinnear, if you agree, could you please merge into the suggestion above?

Author: To be super pedantic, are we saying: 

- when it provides missing functionality, or otherwise eases deployability for
use cases, where there are multiple implementation and/or deployment proponents.

- when it provides missing functionality, otherwise eases deployability for
use cases, or where there are multiple implementation and/or deployment proponents.

- when it provides missing functionality or otherwise eases deployability for
use cases, where there are multiple implementation and/or deployment proponents.

I think the third one is correct comma usage + says (functionality OR deployability) AND (implementation proponents OR deployment proponents)

The second is all ORs everywhere, which was not my reading of the intent, but wouldn't be horrible either.

(updating to have the third bullet in the suggestion, but happy to adjust as necessary

Author: Used bullet three in [8af81ca](https://github.com/ietf-wg-masque/wg-materials/pull/7/commits/8af81ca9873dc5f769a87707af732f24be6455bf)

> for real world use cases. The initial set of extensions will be in support of UDP listening, QUIC-aware proxying, and
> datagram priorities. Extensions that help enable CONNECT-UDP or CONNECT-IP services to coordinate or work well in the

Reviewer 2: ```suggestion
for real world use cases. The initial set of extensions will be in support of UDP listening and QUIC-aware proxying. Additional extensions that help enable CONNECT-UDP or CONNECT-IP services to coordinate or work well in the
```

Author: Nice
